src/databricks/labs/blueprint/wheels.py

```diff
@@ -220,17 +220,26 @@ def __init__(self, installation: Installation, product_info: ProductInfo, *, ver
         self._product_info = product_info
         self._verbose = verbose
 
-    def upload_to_dbfs(self) -> str:
+    def upload_to_dbfs(self, force_dependencies: bool = False) -> str:
         """Uploads the wheel to DBFS location of installation and returns the remote path."""
-        with self._local_wheel.open("rb") as f:
-            return self._installation.upload_dbfs(f"wheels/{self._local_wheel.name}", f)
-
-    def upload_to_wsfs(self) -> str:
+        for wheel in self._local_wheel:
+            if force_dependencies or self._product_info.product_name() in wheel.name:
+                with wheel.open("rb") as f:
+                    remote_wheel = self._installation.upload_dbfs(f"wheels/{wheel.name}", f)
+                    if self._product_info.product_name() in wheel.name:
+                        main_wheel_remote_path = remote_wheel
+        return main_wheel_remote_path
+
+    def upload_to_wsfs(self, force_dependencies: bool = False) -> str:
         """Uploads the wheel to WSFS location of installation and returns the remote path."""
-        with self._local_wheel.open("rb") as f:
-            remote_wheel = self._installation.upload(f"wheels/{self._local_wheel.name}", f.read())
-            self._installation.save(Version(self._product_info.version(), remote_wheel, self._now_iso()))
-            return remote_wheel
+        for wheel in self._local_wheel:
+            if force_dependencies or self._product_info.product_name() in wheel.name:
```

Comment:
> slightly modify the _build_wheel to conditionally pass --no-deps flag, e.g. it's explicit _build_wheel(no_deps=True) for existing functionality and _build_wheel(no_deps=False) for this increment

Thanks, all the items above make sense and will be actioned. Just regarding the modification in `_build_wheel`, since it is called in `WheelsV2.__enter__` , it is not straightforward how to pass `no_deps=False` while running `with self._wheels:` in UCX code [#L527C1-L529C49](https://github.com/databrickslabs/ucx/blob/9c67520c3a32e105b1acd85210cc1254a7de262a/src/databricks/labs/ucx/installer/workflows.py#L527C1-L529C49). Any suggestion on the workaround? If I haven't mistaken couple of changes are required in [WorkspaceInstaller](https://github.com/databrickslabs/ucx/blob/9c67520c3a32e105b1acd85210cc1254a7de262a/src/databricks/labs/ucx/install.py#L175) and [GlobalContext](https://github.com/databrickslabs/ucx/blob/9c67520c3a32e105b1acd85210cc1254a7de262a/src/databricks/labs/ucx/contexts/application.py#L333-L334) accordingly.

The changes to add `no_deps` in `_build_wheel` would be:
```python
    def __init__(self, installation: Installation, product_info: ProductInfo, *, verbose: bool = False, no_deps: bool = True):
       ...
        self._no_deps = no_deps
``` 
```python
     def __enter__(self) -> "WheelsV2":
       ...
        self._local_wheel = self._build_wheel(self._tmp_dir.name, verbose=self._verbose, no_deps=self._no_deps)
        return self
``` 
```python
     def _build_wheel(self, tmp_dir: str, *, verbose: bool = False, no_deps: bool = True) -> list[Path]: ...
```

Reply:
Build wheel is a small change:
- Add keyed argument, default value keeps the current behavior 
- it should return a list of paths instead of a single one 


The rest of the increment is in new methods, that call build wheel for the second time.